src/player.cpp

```diff
@@ -10793,65 +10790,67 @@ bool player::takeoff( item *target, bool autodrop, std::vector<item> *items)
-                            j != worn_index) {
-                        if( autodrop || items != nullptr ) {
-                            if( items != nullptr ) {
-                                items->push_back( worn[j] );
-                            } else {
-                                g->m.add_item_or_charges( posx(), posy(), worn[j] );
-                            }
-                            add_msg(_("You take off your %s."), worn[j].tname().c_str());
-                            worn.erase(worn.begin() + j);
-                            // If we are before worn_index, erasing this element shifted its position by 1.
-                            if (worn_index > j) {
-                                worn_index -= 1;
-                                w = worn[worn_index];
-                            }
-                            taken_off = true;
-                        } else {
-                            add_msg(m_info, _("You can't take off power armor while wearing other power armor components."));
-                            return false;
-                        }
-                    }
-                }
+        // TODO: is this case even used anymore?
+        return wield( nullptr, autodrop );
+    }
+
+    int worn_index = worn_position_to_index( inventory_position );
+    if( static_cast<size_t>( worn_index ) >= worn.size() ) {
+        add_msg( m_info, _("You are not wearing that item.") );
+        return false;
+    }
+    bool taken_off = false;
+
+    auto first_iter = worn.begin();
+    std::advance( first_iter, worn_index );
+    item &w = *first_iter;
+
+    // Handle power armor.
+    if (w.is_power_armor() && w.covers(bp_torso)) {
+        // We're trying to take off power armor, but cannot do that if we have a power armor component on!
+        for( auto iter = worn.begin(); iter != worn.end(); ) {
+            item& other_armor = *iter;
+
+            if( &other_armor == &w || !other_armor.is_power_armor() ) {
+                ++iter;
+                continue;
+            }
+            if( !autodrop && items == nullptr ) {
+                add_msg( m_info, _("You can't take off power armor while wearing other power armor components.") );
+                return false;
             }
 
             if( items != nullptr ) {
-                items->push_back( w );
-                taken_off = true;
-            } else if (autodrop || volume_capacity() - w.get_storage() > volume_carried() + w.volume()) {
-                inv.add_item_keep_invlet(w);
-                taken_off = true;
-            } else if (query_yn(_("No room in inventory for your %s.  Drop it?"),
-                    w.tname().c_str())) {
-                g->m.add_item_or_charges(posx(), posy(), w);
-                taken_off = true;
+                items->push_back( other_armor );
             } else {
-                taken_off = false;
+                g->m.add_item_or_charges( posx(), posy(), other_armor );
```

Comment:
`pos()` would be better than `posx(), posy()`. Currently doesn't matter, but may in the future.